internal/app/export/service.go

```diff
@@ -4,62 +4,248 @@
-	Mqtt   mqtt.Client
-	Logger log.Logger
-	Cfg    config.Config
-	Routes []routes.Route
+	ID        string
+	Mqtt      mqtt.Client
+	Logger    logger.Logger
+	Cfg       config.Config
+	Consumers map[string]routes.Route
+	Cache     messages.Cache
+	connected chan bool
+	status    uint32
+	sync.RWMutex
 }
 
+const (
+	exportGroup = "export-group"
+	count       = 100
+
+	disconnected uint32 = iota
+	connected
+)
+
 // New create new instance of export service
-func New(mqtt mqtt.Client, c config.Config, logger log.Logger) Service {
-	routes := make([]routes.Route, 0)
+func New(c config.Config, cache messages.Cache, l logger.Logger) Service {
+	routes := make(map[string]routes.Route)
+	id := fmt.Sprintf("export-%s", c.MQTT.Username)
+
 	e := exporter{
-		Mqtt:   mqtt,
-		Logger: logger,
-		Cfg:    c,
-		Routes: routes,
+		ID:        id,
+		Logger:    l,
+		Cfg:       c,
+		Consumers: routes,
+		Cache:     cache,
+		connected: make(chan bool, 1),
 	}
-
+	client, err := e.mqttConnect(c, l)
+	if err != nil {
+		log.Fatalf(err.Error())
+	}
+	e.Mqtt = client
 	return &e
 }
 
-// Start method starts consuming messages received from NATS
-// and makes routes according to the configuration file.
-// Routes export messages to mqtt.
-func (e *exporter) Start(queue string, nc *nats.Conn) {
+// Start method loads route configuration
+func (e *exporter) Start(queue string) {
 	var route routes.Route
 	for _, r := range e.Cfg.Routes {
 		switch r.Type {
 		case "mfx":
-			route = mfx.NewRoute(r.NatsTopic, r.MqttTopic, r.SubTopic, e.Mqtt, e.Logger)
+			route = mfx.NewRoute(r.NatsTopic, r.MqttTopic, r.SubTopic)
 		default:
-			route = routes.NewRoute(r.NatsTopic, r.MqttTopic, r.SubTopic, e.Mqtt, e.Logger)
+			route = routes.NewRoute(r.NatsTopic, r.MqttTopic, r.SubTopic)
 		}
-		e.Routes = append(e.Routes, route)
-		e.Subscribe(route, queue, nc)
+
+		if e.validateTopic(route.NatsTopic()) {
+			e.Consumers[route.NatsTopic()] = route
+			g, err := e.Cache.GroupCreate(r.NatsTopic, exportGroup)
+			if err != nil {
+				e.Logger.Error(fmt.Sprintf("Failed to create stream group %s", err.Error()))
+			}
+			e.Logger.Info(fmt.Sprintf("Stream group %s created %s", route.NatsTopic(), g))
+		}
+
 	}
+	e.startRepublish()
 }
 
-func (e *exporter) Subscribe(r routes.Route, queue string, nc *nats.Conn) {
-	_, err := nc.QueueSubscribe(r.NatsTopic(), fmt.Sprintf("%s-%s", queue, r.NatsTopic()), r.Consume)
+func (e *exporter) Consume(msg *nats.Msg) {
+	if _, ok := e.Consumers[msg.Subject]; !ok {
+		e.Logger.Info(fmt.Sprintf("no configuration for nats topic %s", msg.Subject))
+		return
+	}
+
+	route := e.Consumers[msg.Subject]
+	payload, err := route.Consume(msg)
 	if err != nil {
-		e.Logger.Error(fmt.Sprintf("Failed to subscribe for NATS/MQTT %s/%s", r.NatsTopic(), r.MqttTopic()))
+		e.Logger.Error(fmt.Sprintf("Failed to consume msg %s", err.Error()))
+	}
+
+	if err = e.publish(route.MqttTopic(), payload); err != nil {
+		_, err := e.Cache.Add(msg.Subject, route.MqttTopic(), payload)
```

Comment:
Please add a comment here, because it is unclear why we are doing something with cache in the case of error (typically reader would except only logging error and exiting)